# `Multi - index`

In [1]:
import numpy as np
import pandas as pd

### `A multi index object is an index object with multiple levels.`

- To create a multi index, we can pass multiple columns inside the `index_col` parameter while we are importing a csv file.

In [2]:
bigmac_url = 'https://raw.githubusercontent.com/sameerjha462000/datasets/main/bigmac.csv'
bigmac = pd.read_csv(bigmac_url, index_col=['Date', 'Country'])
bigmac

Price in US Dollars
Date       Country                           
2000-04-01 Argentina                 2.500000
           Australia                 1.541667
           Brazil                    1.648045
           Canada                    1.938776
           Switzerland               3.470588
...                                       ...
2020-07-01 Ukraine                   2.174714
           Uruguay                   4.327418
           United States             5.710000
           Vietnam                   2.847282
           South Africa              1.859349

[1386 rows x 1 columns]

In [3]:
type(bigmac.index) # This is a multi index

pandas.core.indexes.multi.MultiIndex

- We can also create a multi index object by passing multiple column names inside the set_index() method.

In [4]:
bigmac_url = 'https://raw.githubusercontent.com/sameerjha462000/datasets/main/bigmac.csv'
bigmac = pd.read_csv(bigmac_url)
bigmac

Date        Country  Price in US Dollars
0     2000-04-01      Argentina             2.500000
1     2000-04-01      Australia             1.541667
2     2000-04-01         Brazil             1.648045
3     2000-04-01         Canada             1.938776
4     2000-04-01    Switzerland             3.470588
...          ...            ...                  ...
1381  2020-07-01        Ukraine             2.174714
1382  2020-07-01        Uruguay             4.327418
1383  2020-07-01  United States             5.710000
1384  2020-07-01        Vietnam             2.847282
1385  2020-07-01   South Africa             1.859349

[1386 rows x 3 columns]

In [5]:
bigmac.set_index(keys = ['Date', 'Country'])

Price in US Dollars
Date       Country                           
2000-04-01 Argentina                 2.500000
           Australia                 1.541667
           Brazil                    1.648045
           Canada                    1.938776
           Switzerland               3.470588
...                                       ...
2020-07-01 Ukraine                   2.174714
           Uruguay                   4.327418
           United States             5.710000
           Vietnam                   2.847282
           South Africa              1.859349

[1386 rows x 1 columns]

# `creating multi index object`

In [6]:
names = ['Abhishek', 'Daniyaal']
subjects = ['Phy', 'Chem', 'Math']

m1 = pd.MultiIndex.from_product([names, subjects])
m1

MultiIndex([('Abhishek',  'Phy'),
            ('Abhishek', 'Chem'),
            ('Abhishek', 'Math'),
            ('Daniyaal',  'Phy'),
            ('Daniyaal', 'Chem'),
            ('Daniyaal', 'Math')],
           )

In [7]:
index_values = []

for name in names:
    for subject in subjects:
        index_values.append((name, subject))
        
index_values

[('Abhishek', 'Phy'),
 ('Abhishek', 'Chem'),
 ('Abhishek', 'Math'),
 ('Daniyaal', 'Phy'),
 ('Daniyaal', 'Chem'),
 ('Daniyaal', 'Math')]

In [8]:
m2 = pd.MultiIndex.from_tuples(index_values)
m2

MultiIndex([('Abhishek',  'Phy'),
            ('Abhishek', 'Chem'),
            ('Abhishek', 'Math'),
            ('Daniyaal',  'Phy'),
            ('Daniyaal', 'Chem'),
            ('Daniyaal', 'Math')],
           )

In [9]:
m1.__class__.__name__, m2.__class__.__name__

('MultiIndex', 'MultiIndex')

## `levels inside multi index object`

In [10]:
names = ['Abhishek', 'Daniyaal']
subjects = ['Phy', 'Chem', 'Math']

m1 = pd.MultiIndex.from_product([names, subjects])
m1

MultiIndex([('Abhishek',  'Phy'),
            ('Abhishek', 'Chem'),
            ('Abhishek', 'Math'),
            ('Daniyaal',  'Phy'),
            ('Daniyaal', 'Chem'),
            ('Daniyaal', 'Math')],
           )

In [11]:
m1.levels

FrozenList([['Abhishek', 'Daniyaal'], ['Chem', 'Math', 'Phy']])

In [12]:
m1.levels[0]

Index(['Abhishek', 'Daniyaal'], dtype='object')

## `stack() and unstack()`

    The stack() method is used to move the column index to row index.
    
    The unstack() method does the reverse i.e moves row index to column index.

In [13]:
index_val = [('cse',2019),('cse',2020),('cse',2021),('cse',2022),('ece',2019),('ece',2020),('ece',2021),('ece',2022)]
multi_index = pd.MultiIndex.from_tuples(index_val)
multi_index

MultiIndex([('cse', 2019),
            ('cse', 2020),
            ('cse', 2021),
            ('cse', 2022),
            ('ece', 2019),
            ('ece', 2020),
            ('ece', 2021),
            ('ece', 2022)],
           )

In [14]:
s = pd.Series([1,2,3,4,5,6,7,8],index=multi_index)
s

cse  2019    1
     2020    2
     2021    3
     2022    4
ece  2019    5
     2020    6
     2021    7
     2022    8
dtype: int64

In [15]:
temp = s.unstack() # moved row index to column index
temp

2019  2020  2021  2022
cse     1     2     3     4
ece     5     6     7     8

In [16]:
temp.stack() # moves column index to row index

cse  2019    1
     2020    2
     2021    3
     2022    4
ece  2019    5
     2020    6
     2021    7
     2022    8
dtype: int64

In [17]:
branch_df1 = pd.DataFrame(
    [
        [1,2],
        [3,4],
        [5,6],
        [7,8],
        [9,10],
        [11,12],
        [13,14],
        [15,16],
    ],
    index = multi_index,
    columns = ['avg_package','students']
)

branch_df1

avg_package  students
cse 2019            1         2
    2020            3         4
    2021            5         6
    2022            7         8
ece 2019            9        10
    2020           11        12
    2021           13        14
    2022           15        16

In [18]:
temp = branch_df1.unstack()
temp

avg_package                students               
           2019 2020 2021 2022     2019 2020 2021 2022
cse           1    3    5    7        2    4    6    8
ece           9   11   13   15       10   12   14   16

In [19]:
temp.unstack() # moves row index further to column index

avg_package  2019  cse     1
                   ece     9
             2020  cse     3
                   ece    11
             2021  cse     5
                   ece    13
             2022  cse     7
                   ece    15
students     2019  cse     2
                   ece    10
             2020  cse     4
                   ece    12
             2021  cse     6
                   ece    14
             2022  cse     8
                   ece    16
dtype: int64

In [20]:
branch_df2 = pd.DataFrame(
    [
        [1,2,0,0],
        [3,4,0,0],
        [5,6,0,0],
        [7,8,0,0],
    ],
    index = [2019,2020,2021,2022],
    columns = pd.MultiIndex.from_product([['delhi','mumbai'],['avg_package','students']])
)

branch_df2

delhi               mumbai         
     avg_package students avg_package students
2019           1        2           0        0
2020           3        4           0        0
2021           5        6           0        0
2022           7        8           0        0

In [21]:
temp = branch_df2.stack() # The stack() method is used to move column index to row index
temp

delhi  mumbai
2019 avg_package      1       0
     students         2       0
2020 avg_package      3       0
     students         4       0
2021 avg_package      5       0
     students         6       0
2022 avg_package      7       0
     students         8       0

In [22]:
temp.stack() # does not make sense however

2019  avg_package  delhi     1
                   mumbai    0
      students     delhi     2
                   mumbai    0
2020  avg_package  delhi     3
                   mumbai    0
      students     delhi     4
                   mumbai    0
2021  avg_package  delhi     5
                   mumbai    0
      students     delhi     6
                   mumbai    0
2022  avg_package  delhi     7
                   mumbai    0
      students     delhi     8
                   mumbai    0
dtype: int64

In [23]:
branch_df3 = pd.DataFrame(
    [
        [1,2,0,0],
        [3,4,0,0],
        [5,6,0,0],
        [7,8,0,0],
        [9,10,0,0],
        [11,12,0,0],
        [13,14,0,0],
        [15,16,0,0],
    ],
    index = multi_index,
    columns = pd.MultiIndex.from_product([['delhi','mumbai'],['avg_package','students']])
)

branch_df3

delhi               mumbai         
         avg_package students avg_package students
cse 2019           1        2           0        0
    2020           3        4           0        0
    2021           5        6           0        0
    2022           7        8           0        0
ece 2019           9       10           0        0
    2020          11       12           0        0
    2021          13       14           0        0
    2022          15       16           0        0

In [24]:
temp = branch_df3.unstack()
temp

delhi                                             mumbai            \
    avg_package                students                avg_package             
           2019 2020 2021 2022     2019 2020 2021 2022        2019 2020 2021   
cse           1    3    5    7        2    4    6    8           0    0    0   
ece           9   11   13   15       10   12   14   16           0    0    0   

                                  
         students                 
    2022     2019 2020 2021 2022  
cse    0        0    0    0    0  
ece    0        0    0    0    0

In [25]:
temp.unstack() # does not make sense

delhi   avg_package  2019  cse     1
                           ece     9
                     2020  cse     3
                           ece    11
                     2021  cse     5
                           ece    13
                     2022  cse     7
                           ece    15
        students     2019  cse     2
                           ece    10
                     2020  cse     4
                           ece    12
                     2021  cse     6
                           ece    14
                     2022  cse     8
                           ece    16
mumbai  avg_package  2019  cse     0
                           ece     0
                     2020  cse     0
                           ece     0
                     2021  cse     0
                           ece     0
                     2022  cse     0
                           ece     0
        students     2019  cse     0
                           ece     0
                     2020  cse     0
 

In [26]:
temp.stack()

delhi               mumbai         
         avg_package students avg_package students
cse 2019           1        2           0        0
    2020           3        4           0        0
    2021           5        6           0        0
    2022           7        8           0        0
ece 2019           9       10           0        0
    2020          11       12           0        0
    2021          13       14           0        0
    2022          15       16           0        0

In [27]:
temp.stack().stack()

delhi  mumbai
cse 2019 avg_package      1       0
         students         2       0
    2020 avg_package      3       0
         students         4       0
    2021 avg_package      5       0
         students         6       0
    2022 avg_package      7       0
         students         8       0
ece 2019 avg_package      9       0
         students        10       0
    2020 avg_package     11       0
         students        12       0
    2021 avg_package     13       0
         students        14       0
    2022 avg_package     15       0
         students        16       0

# `working with MultiIndex dataframes`

In [28]:
branch_df3

delhi               mumbai         
         avg_package students avg_package students
cse 2019           1        2           0        0
    2020           3        4           0        0
    2021           5        6           0        0
    2022           7        8           0        0
ece 2019           9       10           0        0
    2020          11       12           0        0
    2021          13       14           0        0
    2022          15       16           0        0

In [29]:
branch_df3.head()

delhi               mumbai         
         avg_package students avg_package students
cse 2019           1        2           0        0
    2020           3        4           0        0
    2021           5        6           0        0
    2022           7        8           0        0
ece 2019           9       10           0        0

In [30]:
branch_df3.tail()

delhi               mumbai         
         avg_package students avg_package students
cse 2022           7        8           0        0
ece 2019           9       10           0        0
    2020          11       12           0        0
    2021          13       14           0        0
    2022          15       16           0        0

In [31]:
branch_df3.sample(5)

delhi               mumbai         
         avg_package students avg_package students
ece 2019           9       10           0        0
cse 2019           1        2           0        0
    2022           7        8           0        0
ece 2022          15       16           0        0
cse 2021           5        6           0        0

In [32]:
branch_df3

delhi               mumbai         
         avg_package students avg_package students
cse 2019           1        2           0        0
    2020           3        4           0        0
    2021           5        6           0        0
    2022           7        8           0        0
ece 2019           9       10           0        0
    2020          11       12           0        0
    2021          13       14           0        0
    2022          15       16           0        0

In [33]:
branch_df3.sort_values(by = ('delhi', 'avg_package'), ascending = False)

delhi               mumbai         
         avg_package students avg_package students
ece 2022          15       16           0        0
    2021          13       14           0        0
    2020          11       12           0        0
    2019           9       10           0        0
cse 2022           7        8           0        0
    2021           5        6           0        0
    2020           3        4           0        0
    2019           1        2           0        0

In [34]:
branch_df3.isnull().sum()

delhi   avg_package    0
        students       0
mumbai  avg_package    0
        students       0
dtype: int64

In [35]:
branch_df3.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 8 entries, ('cse', 2019) to ('ece', 2022)
Data columns (total 4 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   (delhi, avg_package)   8 non-null      int64
 1   (delhi, students)      8 non-null      int64
 2   (mumbai, avg_package)  8 non-null      int64
 3   (mumbai, students)     8 non-null      int64
dtypes: int64(4)
memory usage: 632.0+ bytes


In [36]:
branch_df3.describe()

delhi                 mumbai         
      avg_package   students avg_package students
count    8.000000   8.000000         8.0      8.0
mean     8.000000   9.000000         0.0      0.0
std      4.898979   4.898979         0.0      0.0
min      1.000000   2.000000         0.0      0.0
25%      4.500000   5.500000         0.0      0.0
50%      8.000000   9.000000         0.0      0.0
75%     11.500000  12.500000         0.0      0.0
max     15.000000  16.000000         0.0      0.0

# `Accessing rows and columns`

In [37]:
branch_df3

delhi               mumbai         
         avg_package students avg_package students
cse 2019           1        2           0        0
    2020           3        4           0        0
    2021           5        6           0        0
    2022           7        8           0        0
ece 2019           9       10           0        0
    2020          11       12           0        0
    2021          13       14           0        0
    2022          15       16           0        0

In [38]:
branch_df3['delhi']

avg_package  students
cse 2019            1         2
    2020            3         4
    2021            5         6
    2022            7         8
ece 2019            9        10
    2020           11        12
    2021           13        14
    2022           15        16

In [39]:
branch_df3[('delhi', 'avg_package')]

cse  2019     1
     2020     3
     2021     5
     2022     7
ece  2019     9
     2020    11
     2021    13
     2022    15
Name: (delhi, avg_package), dtype: int64

In [40]:
branch_df3.head()

delhi               mumbai         
         avg_package students avg_package students
cse 2019           1        2           0        0
    2020           3        4           0        0
    2021           5        6           0        0
    2022           7        8           0        0
ece 2019           9       10           0        0

In [41]:
branch_df3.iloc[1]

delhi   avg_package    3
        students       4
mumbai  avg_package    0
        students       0
Name: (cse, 2020), dtype: int64

In [42]:
branch_df3.iloc[1:4]

delhi               mumbai         
         avg_package students avg_package students
cse 2020           3        4           0        0
    2021           5        6           0        0
    2022           7        8           0        0

In [43]:
branch_df3.iloc[[1,4,3]]

delhi               mumbai         
         avg_package students avg_package students
cse 2020           3        4           0        0
ece 2019           9       10           0        0
cse 2022           7        8           0        0

In [44]:
branch_df3.iloc[[1,4,3], 1::]

delhi      mumbai         
         students avg_package students
cse 2020        4           0        0
ece 2019       10           0        0
cse 2022        8           0        0

In [45]:
branch_df3.iloc[[1,4,3], ::2]

delhi      mumbai
         avg_package avg_package
cse 2020           3           0
ece 2019           9           0
cse 2022           7           0

In [46]:
branch_df3.iloc[[1,4,3], [0, 3, 1]]

delhi   mumbai    delhi
         avg_package students students
cse 2020           3        0        4
ece 2019           9        0       10
cse 2022           7        0        8

### `The loc accessor also works the same way as iloc accessor`

In [47]:
branch_df3

delhi               mumbai         
         avg_package students avg_package students
cse 2019           1        2           0        0
    2020           3        4           0        0
    2021           5        6           0        0
    2022           7        8           0        0
ece 2019           9       10           0        0
    2020          11       12           0        0
    2021          13       14           0        0
    2022          15       16           0        0

In [48]:
branch_df3.loc[('cse', 2019)]

delhi   avg_package    1
        students       2
mumbai  avg_package    0
        students       0
Name: (cse, 2019), dtype: int64

In [49]:
branch_df3.loc[('cse', 2019) : ('ece', 2021) : 2]

delhi               mumbai         
         avg_package students avg_package students
cse 2019           1        2           0        0
    2021           5        6           0        0
ece 2019           9       10           0        0
    2021          13       14           0        0

In [50]:
branch_df3.loc[[('cse', 2020), ('ece', 2022), ('cse', 2019)]]

delhi               mumbai         
         avg_package students avg_package students
cse 2020           3        4           0        0
ece 2022          15       16           0        0
cse 2019           1        2           0        0

In [51]:
branch_df3.loc[('cse', 2019), [('delhi', 'avg_package'), ('mumbai', 'avg_package')]]

delhi   avg_package    1
mumbai  avg_package    0
Name: (cse, 2019), dtype: int64

In [52]:
branch_df3.loc[('cse', 2019) : ('ece', 2021) : 2, ('delhi', 'avg_package'): :3]

delhi   mumbai
         avg_package students
cse 2019           1        0
    2021           5        0
ece 2019           9        0
    2021          13        0

In [53]:
branch_df3.loc[[('cse', 2020), ('ece', 2022), ('cse', 2019)], [('delhi', 'avg_package'), ('mumbai', 'avg_package')]]

delhi      mumbai
         avg_package avg_package
cse 2020           3           0
ece 2022          15           0
cse 2019           1           0

# `swaplevel`

In [54]:
df = pd.DataFrame(
    {"Grade": ["A", "B", "A", "C"]},
    index=[
        ["Final exam", "Final exam", "Coursework", "Coursework"],
        ["History", "Geography", "History", "Geography"],
        ["January", "February", "March", "April"],
    ],
)
df

Grade
Final exam History   January      A
           Geography February     B
Coursework History   March        A
           Geography April        C

    In the following example, we will swap the levels of the indices. Here, we will swap the levels column-wise, but levels can be swapped row-wise in a similar manner. Note that column-wise is the default behaviour. By not supplying any arguments for i and j, we swap the last and second to last indices.

In [55]:
df.swaplevel()

Grade
Final exam January  History       A
           February Geography     B
Coursework March    History       A
           April    Geography     C

    By supplying one argument, we can choose which index to swap the last index with. We can for example swap the first index with the last one as follows.

In [56]:
df.swaplevel(0)

Grade
January  History   Final exam     A
February Geography Final exam     B
March    History   Coursework     A
April    Geography Coursework     C

    We can also define explicitly which indices we want to swap by supplying values for both i and j. Here, we for example swap the first and second indices.

In [57]:
df.swaplevel(0, 1)

Grade
History   Final exam January      A
Geography Final exam February     B
History   Coursework March        A
Geography Coursework April        C

# `Long and Wide data`

    These two are different forms of storing tabular data. We may need to work with any of these depending on the situation.
    Hence we must know when and how to use them and convert them into one another.
    
    Long data is a data format in which we have a dedicated row for each attribute of a data point.
    
    Wide data is a data format in which we have a dedicated column for each attribute of a data point.
    
    melt() method is used to convert a wide data format into a long data format.

In [64]:
df = pd.DataFrame({
    'name':['Abhishek', 'Ibrahim', 'Daniyaal'],
    'age' : [23, 23, 45],
    'weight':[75, 67, 89],
    'height':[6, 5.99, 5.8]
})
df

name  age  weight  height
0  Abhishek   23      75    6.00
1   Ibrahim   23      67    5.99
2  Daniyaal   45      89    5.80

In [67]:
df.melt(id_vars='name', var_name = 'attribute', value_name = 'value')

name attribute  value
0  Abhishek       age  23.00
1   Ibrahim       age  23.00
2  Daniyaal       age  45.00
3  Abhishek    weight  75.00
4   Ibrahim    weight  67.00
5  Daniyaal    weight  89.00
6  Abhishek    height   6.00
7   Ibrahim    height   5.99
8  Daniyaal    height   5.80

## `Melting real life data`

In [68]:
deaths = pd.read_csv('time_series_covid19_deaths_global.csv')
deaths.head()

Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  ...  12/24/22  12/25/22  12/26/22  \
0        0        0        0        0  ...      7845      7846      7846   
1        0        0        0        0  ...      3595      3595      3595   
2        0        0        0        0  ...      6881      6881      6881   
3        0        0        0        0  ...       165       165       165   
4        0        0        0        0  ...      1928      1928      1928   

   12/27/22  12/28/22  12/29/22  12/30/22  12/31/22  1/1/23  1/2/23  
0      7846      7846      7847      7847      7849    7849    7849  
1      3595      3595      3595      3595      3595    3595    3595  
2      6881      6881      6881      6881      6881    6881    6881  
3       165       165       165       165       165     165     165  
4      1930      1930      1930      1930      1930    1930    1930  

[5 rows x 1081 columns]

`Currently we have 1081 rows which is very difficult to deal with. Hence we will melt this dataframe.`

In [69]:
deaths.melt(id_vars = ['Province/State', 'Country/Region', 'Lat', 'Long'], vard_name = 'Date', value_name = 'num_deaths')

Province/State        Country/Region        Lat        Long     Date  \
0                 NaN           Afghanistan  33.939110   67.709953  1/22/20   
1                 NaN               Albania  41.153300   20.168300  1/22/20   
2                 NaN               Algeria  28.033900    1.659600  1/22/20   
3                 NaN               Andorra  42.506300    1.521800  1/22/20   
4                 NaN                Angola -11.202700   17.873900  1/22/20   
...               ...                   ...        ...         ...      ...   
311248            NaN    West Bank and Gaza  31.952200   35.233200   1/2/23   
311249            NaN  Winter Olympics 2022  39.904200  116.407400   1/2/23   
311250            NaN                 Yemen  15.552727   48.516388   1/2/23   
311251            NaN                Zambia -13.133897   27.849332   1/2/23   
311252            NaN              Zimbabwe -19.015438   29.154857   1/2/23   

        num_deaths  
0                0  
1                0  
2                0  
3                0  
4                0  
...            ...  
311248        5708  
311249           0  
311250        2159  
311251        4024  
311252        5637  

[311253 rows x 6 columns]

In [70]:
confirm = pd.read_csv('time_series_covid19_confirmed_global.csv')
confirm.head()

Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  ...  12/24/22  12/25/22  12/26/22  \
0        0        0        0        0  ...    207310    207399    207438   
1        0        0        0        0  ...    333749    333749    333751   
2        0        0        0        0  ...    271194    271198    271198   
3        0        0        0        0  ...     47686     47686     47686   
4        0        0        0        0  ...    104973    104973    104973   

   12/27/22  12/28/22  12/29/22  12/30/22  12/31/22  1/1/23  1/2/23  
0    207460    207493    207511    207550    207559  207616  207627  
1    333751    333776    333776    333806    333806  333811  333812  
2    271202    271208    271217    271223    271228  271229  271229  
3     47686     47751     47751     47751     47751   47751   47751  
4    105095    105095    105095    105095    105095  105095  105095  

[5 rows x 1081 columns]

In [71]:
confirm.melt(id_vars = ['Province/State', 'Country/Region', 'Lat', 'Long'], var_name = 'Date', value_name = 'num_cases')

Province/State        Country/Region        Lat        Long     Date  \
0                 NaN           Afghanistan  33.939110   67.709953  1/22/20   
1                 NaN               Albania  41.153300   20.168300  1/22/20   
2                 NaN               Algeria  28.033900    1.659600  1/22/20   
3                 NaN               Andorra  42.506300    1.521800  1/22/20   
4                 NaN                Angola -11.202700   17.873900  1/22/20   
...               ...                   ...        ...         ...      ...   
311248            NaN    West Bank and Gaza  31.952200   35.233200   1/2/23   
311249            NaN  Winter Olympics 2022  39.904200  116.407400   1/2/23   
311250            NaN                 Yemen  15.552727   48.516388   1/2/23   
311251            NaN                Zambia -13.133897   27.849332   1/2/23   
311252            NaN              Zimbabwe -19.015438   29.154857   1/2/23   

        num_cases  
0               0  
1               0  
2               0  
3               0  
4               0  
...           ...  
311248     703228  
311249        535  
311250      11945  
311251     334661  
311252     259981  

[311253 rows x 6 columns]